Feature/firebase security rules

raubreysmith opened the pull request:
# JIRA
[Farm-109](https://amidodevelopment.atlassian.net/browse/FARM-109)

### :eyeglasses: Review 
@maribel4amido @archiebw-amido 

# DETAILS
### :construction: Change
Adds firebase security rules into the repository with examples for unit testing and config for deployment. Please refer to the updates to the README.md

Before we can merge this we need to confirm our actual rule definitions. We should also consider if we are happy for these to exist in our a public repository, or if we feel that the rules may expose too much of our repository architecture then we could extract this to a separate admin repo.

### :traffic_light: Testing
Unit tests are included in the repository, however deployed rules can be manually tested using the simulator in the Firebase console - links in the readme doc

### :heavy_check_mark: Coding Standard Checklist

- [x] Unit Test Coverage 80%
- [x] TODO tasks has corresponding JIRA IDs if not completed in this PR
- [ ] Smoke Test done or deployed on Firebase
- [ ] Documentation provided: Systems Design in Confluence, Configuration Guide, Installation Guide, Runbook where applicable.

maribel4amido commented on firebase.json:
```diff
@@ -1,7 +1,14 @@
 {
+  "database": {
+    "rules": "rules/database/database.rules.json"
```
I see that Realtime database rules and tests have been introduced but I do question whether we should have these rules and tests since we are not using Realtime database in any of our clients. I think that having this in code makes the repo more complex (due to all files and tests to do) than it is and gives an impression that we're using Realtime database which we are not.

raubreysmith replied:
true, I had considered that it would be useful to include the rules in our code to ensure that no one gave access through the console. By defining them here it is a single place for all the definitions.

raubreysmith replied:
remove unused rules and tests

maribel4amido commented on rules/firestore/authenticated.test.js:
```diff
@@ -0,0 +1,46 @@
+let app;
+
+// Firestore emulator must be running
+// firebase serve --only firestore
```
In the context of CI, is it recommended to run firebase serve? 
By that I mean, is there a way to start firebase serve, run tests (and regardless of result), kill the process of firebase serve?
I'd be interested to see how this behaves on Circle CI

raubreysmith replied:
this needs some further investigation, I was just looking at our CI config as well and it could be that keeping all the rules separate with their own NPM package creates unnecessary work on the server. As for the emulation, I'm doing some further investigation as it looks like it [might not be straight forward on Circle](https://github.com/firebase/quickstart-nodejs/issues/48)

maribel4amido commented on rules/firestore/firestore.rules:
```diff
@@ -1,7 +1,7 @@
   match /databases/{database}/documents {
     match /{document=**} {
-      allow read, write: if false;
+      allow read, write: if request.auth.uid != null;
```
It'd be good to also have a test for a rule where only a specific user is allowed read or write so that another authenticated user can't read or write the secured document.

raubreysmith replied:
easily done, I'll provide a few additional examples

raubreysmith replied:
added user_access.test.js with examples of accessing collections under a uid.

raubreysmith commented:
@maribel4amido ok I've changed the ruleset and added a more complete set of tests. You can add a verbose flag `npm test -- --verbose=true` to see them all - I've split the texts by context then the bulk are dynamically run using collections defined in `collections.js`
